internal/ingress/backend/endpoint.go

```diff
@@ -93,14 +146,72 @@ func (resolver *endpointResolver) resolveIP(ingress *extensions.Ingress, backend
 		return nil, fmt.Errorf("Unable to find service endpoints for %s: %v", serviceKey, err.Error())
 	}
 
+	podMap := map[string]*corev1.Pod{}
+	pods := resolver.store.GetServicePods(service.Spec.Selector)
+	for _, pod := range pods {
+		podMap[pod.Name] = pod
+	}
+	conditionTypePrefix := fmt.Sprintf("target-health.%s/", parser.AnnotationsPrefix)
+	conditionType := api.PodConditionType(fmt.Sprintf("%s%s_%s_%s", conditionTypePrefix, ingress.Name, service.Name, backend.ServicePort.String()))
+
 	var result []*elbv2.TargetDescription
 	for _, epSubset := range eps.Subsets {
 		for _, epPort := range epSubset.Ports {
 			// servicePort.Name is optional if there is only one port
 			if servicePort.Name != "" && servicePort.Name != epPort.Name {
 				continue
 			}
-			for _, epAddr := range epSubset.Addresses {
+			addresses := epSubset.Addresses
+			if service.Spec.PublishNotReadyAddresses {
+				addresses = append(addresses, epSubset.NotReadyAddresses...)
+			} else {
+				// if `PublishNotReadyAddresses` is not set, we need to loop over all unready pods to check if the ALB readiness gate is the only condition preventing the pod from being ready;
+				// if this is the case, we return the pod as a desired target although its not in `Addresses`
```

Comment:
this is a hacky logic. If there are another similar controller using same logic (only try to reconcile condition if it's the only pending condition), it will be dead lock..

Reply:
> If there are another similar controller using same logic (only try to reconcile condition if it's the only pending condition), it will be dead-lock..

That's a good point. But what if you have other readiness gates that need to be fulfilled until a pod really gets ready to accept traffic? If you ignore those, the pod will be registered and healthy in the load balancer maybe without a necessary side condition (represented by the other readiness gate) being fulfilled.

Right now, we don't have other readiness gates, but that might change in the near future and these would need to be fulfilled before the pod gets registered. How about for now I change the code to only look at `ContainersReady` and later on we could introduce another annotation like `<prefix>/wait-for-pod-conditions: '["readiness-gate-1", "readiness-gate-2"]'` to wait for specific conditions before registering the pod? Would that make sense to you?

Reply:
yeah, make sense to me :D